sync15-adapter/src/record_types.rs

```diff
@@ -0,0 +1,241 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use bso_record::{BsoRecord, Sync15Record};
+use std::collections::HashMap;
+
+pub use MaybeTombstone::*;
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum MaybeTombstone<T> {
+    Tombstone { id: String, deleted: bool },
+    NonTombstone(T)
+}
+
+impl<T> MaybeTombstone<T> {
+    #[inline]
+    pub fn tombstone<R: Into<String>>(id: R) -> MaybeTombstone<T> {
+        Tombstone { id: id.into(), deleted: true }
+    }
+
+    #[inline]
+    pub fn is_tombstone(&self) -> bool {
+        match self {
+            &NonTombstone(_) => false,
+            _ => true
+        }
+    }
+
+    #[inline]
+    pub fn unwrap(self) -> T {
+        match self {
+            NonTombstone(record) => record,
+            _ => panic!("called `MaybeTombstone::unwrap()` on a Tombstone!"),
+        }
+    }
+
+    #[inline]
+    pub fn expect(self, msg: &str) -> T {
+        match self {
+            NonTombstone(record) => record,
+            _ => panic!("{}", msg),
+        }
+    }
+
+    #[inline]
+    pub fn ok_or<E>(self, err: E) -> ::std::result::Result<T, E> {
+        match self {
+            NonTombstone(record) => Ok(record),
+            _ => Err(err)
+        }
+    }
+
+    #[inline]
+    pub fn record(self) -> Option<T> {
+        match self {
+            NonTombstone(record) => Some(record),
+            _ => None
+        }
+    }
+}
+
+impl<T> Sync15Record for MaybeTombstone<T> where T: Sync15Record {
+    fn collection_tag() -> &'static str { T::collection_tag() }
+    fn ttl() -> Option<u32> { T::ttl() }
+    fn record_id(&self) -> &str {
+        match self {
+            &Tombstone { ref id, .. } => id,
+            &NonTombstone(ref record) => record.record_id()
+        }
+    }
+    fn sortindex(&self) -> Option<i32> {
+        match self {
+            &Tombstone { .. } => None,
+            &NonTombstone(ref record) => record.sortindex()
+        }
+    }
+}
+
+impl<T> BsoRecord<MaybeTombstone<T>> {
+    #[inline]
+    pub fn is_tombstone(&self) -> bool {
+        self.payload.is_tombstone()
+    }
+
+    #[inline]
+    pub fn record(self) -> Option<BsoRecord<T>> where T: Clone {
+        self.map_payload(|payload| payload.record()).transpose()
+    }
+}
+
+pub type MaybeTombstoneRecord<T> = BsoRecord<MaybeTombstone<T>>;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PasswordRecord {
```

Comment:
TBH, I'm not sure that concrete record types belong in this library, but instead closer to the storage of these records. We should consider just exposing decrypted json strings and leave the abstractions to the consumer.

Reply:
I somewhat agree with @mhammond, but only to the extent that I think they should be in a separate file. I think it's better to have canonical implementations than to make consumers figure out the object formats themselves!

It's always possible for the consumer to specify their own `T: Sync15Record` and bypass these.